lints/duplicate-mutable-accounts/src/anchor_constraint.rs

```diff
@@ -0,0 +1,118 @@
+        if let GenericArg::Type(hir_ty) = &generic_args.args[1];
+        then {
+            get_def_id(hir_ty)
+        } else {
+            None
+        }
+    }
+}
+
+/// Returns the `DefId` of `ty`, an hir type. Returns `None` if cannot resolve type.
+pub fn get_def_id(ty: &rustc_hir::Ty) -> Option<DefId> {
+    if_chain! {
+        if let TyKind::Path(qpath) = &ty.kind;
+        if let QPath::Resolved(_, path) = qpath;
+        if let Res::Def(_, def_id) = path.res;
+        then {
+            Some(def_id)
+        } else {
+            None
+        }
+    }
+}
+
+/// Returns a `TokenStream` of form: `a`.key() != `b`.key().
+pub fn create_key_check_constraint_tokenstream(a: Symbol, b: Symbol) -> TokenStream {
+    // TODO: may be more efficient way to do this, since the stream is effectively fixed
+    // and determined. Only two tokens are variable.
+    let constraint = vec![
+        TreeAndSpacing::from(create_token_from_ident(a.as_str())),
+        TreeAndSpacing::from(TokenTree::Token(Token::new(TokenKind::Dot, DUMMY_SP))),
+        TreeAndSpacing::from(create_token_from_ident("key")),
+        TreeAndSpacing::from(TokenTree::Delimited(
+            DelimSpan::dummy(),
+            Delimiter::Parenthesis,
+            TokenStream::new(vec![]),
+        )),
+        TreeAndSpacing::from(TokenTree::Token(Token::new(TokenKind::Ne, DUMMY_SP))),
+        TreeAndSpacing::from(create_token_from_ident(b.as_str())),
+        TreeAndSpacing::from(TokenTree::Token(Token::new(TokenKind::Dot, DUMMY_SP))),
+        TreeAndSpacing::from(create_token_from_ident("key")),
+        TreeAndSpacing::from(TokenTree::Delimited(
+            DelimSpan::dummy(),
+            Delimiter::Parenthesis,
+            TokenStream::new(vec![]),
+        )),
+    ];
+
+    TokenStream::new(constraint)
+}
+
+/// Returns a `TokenTree::Token` which has `TokenKind::Ident`, with the string set to `s`.
+fn create_token_from_ident(s: &str) -> TokenTree {
+    let ident = Ident::from_str(s);
+    TokenTree::Token(Token::from_ast_ident(ident))
+}
+
+#[derive(Debug, Default)]
+pub struct Streams(pub Vec<TokenStream>);
+
+impl Streams {
+    /// Returns true if `self` has a `TokenStream` that `other` is a substream of
+    pub fn contains(&self, other: &TokenStream) -> bool {
+        self.0
+            .iter()
+            .any(|token_stream| Self::is_substream(token_stream, other))
+    }
+
+    /// Returns true if `other` is a substream of `stream`. By substream we mean in the
+    /// sense of a substring.
+    // NOTE: a possible optimization is when a match is found, to remove the matched
+    // TokenTrees from the TokenStream, since the constraint has been "checked" so it never
+    // needs to be validated again. This cuts down the number of comparisons.
+    fn is_substream(stream: &TokenStream, other: &TokenStream) -> bool {
+        let other_len = other.len();
+        for i in 0..stream.len() {
+            for (j, other_token) in other.trees().enumerate() {
```

Comment:
I have a hunch this could be written more simply as:
```rust
if other.trees().enumerate().all(|(j, other_token)| { ... }) {
    return true;
}
```